#! /bin/sh -e
 
# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Fix bugs in Sparc clone implementation.
# DP: Author: David S. Miller <davem@davemloft.net>
# DP: Upstream status: This patch was backported from upstream.
# DP: Status Details:
# DP: Date: 2006-03-21
 
if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p1 < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p1 < $0;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0
 
--- glibc-2.3.6/sysdeps/unix/sysv/linux/sparc/sparc32/clone.S.~1~	2004-12-16 08:46:54.000000000 -0800
+++ glibc-2.3.6/sysdeps/unix/sysv/linux/sparc/sparc32/clone.S	2006-03-21 17:38:37.000000000 -0800
@@ -40,11 +40,19 @@ __clone:
 	save	%sp,-96,%sp
 
 	/* sanity check arguments */
-	tst	%i0
+	orcc	%i0,%g0,%g2
 	be	.Lerror
 	 orcc	%i1,%g0,%o1
 	be	.Lerror
 	 mov	%i2,%o0
+
+       /* The child_stack is the top of the stack, allocate one
+           whole stack frame from that as this is what the kernel
+           expects.  */
+        sub     %o1, 96, %o1
+        mov     %i3, %g3
+        mov     %i2, %g4
+
 	/* ptid */
 	mov	%i4,%o2
 	/* tls */
@@ -59,7 +67,7 @@ __clone:
 	 tst	%o1
 	bne	__thread_start
 	 nop
-	ret
+	jmpl	%i7 + 8, %g0
 	 restore %o0,%g0,%o0
 
 .Lerror:
@@ -76,9 +84,9 @@ __clone:
 __thread_start:
 #ifdef RESET_PID
 	sethi	%hi(CLONE_THREAD), %l0
-	andcc	%i2, %l0, %g0
+	andcc	%g4, %l0, %g0
 	bne	1f
-	 andcc	%i2, CLONE_VM, %g0
+	 andcc	%g4, CLONE_VM, %g0
 	bne,a	2f
 	 mov	-1,%o0
 	set	__NR_getpid,%g1
@@ -87,8 +95,9 @@ __thread_start:
 	st	%o0,[%g7 + TID]
 1:
 #endif
-	call	%i0
-	 mov	%i3,%o0
+	mov	%g0, %fp	/* terminate backtrace */
+	call	%g2
+	 mov	%g3,%o0
 	call	_exit,0
 	 nop
 
--- glibc-2.3.6/sysdeps/unix/sysv/linux/sparc/sparc64/clone.S.~1~	2001-12-18 16:20:53.000000000 -0800
+++ glibc-2.3.6/sysdeps/unix/sysv/linux/sparc/sparc64/clone.S	2006-03-21 17:35:18.000000000 -0800
@@ -25,6 +25,9 @@
 
 /* int clone(int (*fn)(void *arg), void *child_stack, int flags, void *arg); */
 
+	.register	%g2,#scratch
+	.register	%g3,#scratch
+
 	.text
 	.align	4
 	.globl	__clone
@@ -35,12 +38,14 @@ __clone:
 
 	/* sanity check arguments */
 	brz,pn	%i0, 99f
-	 mov	%i0, %l0		/* save fn */
+	 mov	%i0, %g2		/* save fn */
 	brz,pn	%i1, 99f
-	 mov	%i3, %l3		/* save arg */
+	 mov	%i3, %g3		/* save arg */
 
-	/* Do the system call */
-	sub	%i1, 0x7ff, %o1
+        /* The child_stack is the top of the stack, allocate one
+           whole stack frame from that as this is what the kernel
+           expects.  Also, subtract STACK_BIAS.  */
+	sub	%i1, 192 + 0x7ff, %o1
 	mov	%i2, %o0
 	set	__NR_clone, %g1
 	ta	0x6d
@@ -78,9 +83,8 @@ __clone:
 	.type __thread_start,@function
 __thread_start:
 	mov	%g0, %fp	/* terminate backtrace */
-	sub	%sp, 6*8, %sp	/* provide arg storage */
-	call	%l0
-	 mov	%l3,%o0
+	call	%g2
+	 mov	%g3,%o0
 	call	_exit,0
 	 nop
 	.size __thread_start, .-__thread_start
